File: Problem201_shortest-word-distance-iii.java
// Time Complexity: O(n)
// Space Complexity: O(1)
import java.util.*;
class Solution {
    
    public int shortestWordDistance(String[] words, String word1, String word2) {
        HashMap<String, ArrayList<Integer>> locations = new HashMap<String, ArrayList<Integer>>();

        // Prepare a mapping from a word to all it's locations (indices).
        for (int i = 0; i < words.length; i++) {
            ArrayList<Integer> loc = locations.getOrDefault(words[i], new ArrayList<Integer>());
            loc.add(i);
            locations.put(words[i], loc);
        }
        
        ArrayList<Integer> loc1, loc2;

        // Location lists for both the words
        // the indices will be in SORTED order by default
        loc1 = locations.get(word1);
        loc2 = locations.get(word2);

        int l1 = 0, l2 = 0, minDiff = Integer.MAX_VALUE;
        while (l1 < loc1.size() && l2 < loc2.size()) {
            int idx1 = loc1.get(l1);
            int idx2 = loc2.get(l2);
            if(idx1 == idx2) {
                l2++;
                continue;
            }
            minDiff = Math.min(minDiff, Math.abs(idx1 - idx2));
            if (idx1 < idx2) {
                l1++;
            } else {
                l2++;
            }
        }

        return minDiff;
    }
}

File: Problem-201 Shortest Word Distance III.py
# 245. Shortest Word Distance III
# https://leetcode.com/problems/shortest-word-distance-iii/

# Logic:
# Less Efficient: Make a hashmap with the idx of the words and calculate distance.

# Efficient: Have 2 pointers and update them when corresponding word is found. 
# But if both the words are same, w2 becomes w1 everytime a new idx matches.

# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def shortestWordDistance(self, wordsDict: List[str], word1: str, word2: str) -> int:
        w1 = -1
        w2 = -1
        minDist = inf
        
        for idx, i in enumerate(wordsDict):
            if i == word1:
                w1 = idx
            
                if word1 == word2:
                    w1 = w2
            
            if i == word2:
                w2 = idx

            if w1 != -1 and w2 != -1:
                minDist = min(minDist, abs(w1-w2))
        return minDist